### PR TITLE
Speed up backport_linter()

### PR DESCRIPTION
**Pull request by AshesITR**
A slight speedup by using a more efficient data structure for backport checking.

```r
devtools::load_all()
system.time(lint_package(linters = backport_linter("oldrel-4", except = c("R_user_dir", "str2lang", "str2expression", "deparse1", "...names"))))

# here
#    user  system elapsed 
#  5.667   0.011   5.675 

# main
#    user  system elapsed 
#  6.270   0.020   6.286 
```

### CODE REVIEW

**Comment by AshesITR** [comment]
I'll do some more testing to see if I find an even better solution, but avoiding `do.call(rbind, ...)` saves almost 20%.

File: R/backport_linter.R [hunk 1]
```diff
@@ -42,6 +42,9 @@ backport_linter <- function(r_version = getRversion(), except = character()) {
 
   backport_blacklist <- backports[r_version < R_system_version(names(backports))]
   backport_blacklist <- lapply(backport_blacklist, setdiff, except)
+  backport_funs <- unlist(backport_blacklist)
+  backport_versions <- rep(names(backport_blacklist), times = lengths(backport_blacklist))
+  backport_index <- stats::setNames(backport_versions, backport_funs)
```
Comment (MichaelChirico) [inline R/backport_linter.R:47]: why keep `backport_versions` and `backport_index` separately? Why not just `names(backport_versions) <- backport_funs`?

File: NEWS.md [hunk 2]
```diff
@@ -23,6 +23,7 @@
   + to encourage all packages to be attached with `library(symbol)`, not `library("symbol", character.only = TRUE)` or "vectorized" approaches looping over package names (part of #884, @MichaelChirico).
   + to discourage many consecutive calls to `suppressMessages()` or `suppressPackageStartupMessages()` (part of #884, @MichaelChirico).
 * `unnecessary_lambda_linter` is extended to encourage vectorized comparisons where possible, e.g. `sapply(x, sum) > 0` instead of `sapply(x, function(x) sum(x) > 0)` (part of #884, @MichaelChirico). Toggle this behavior with argument `allow_comparison`.
+* `backport_linter()` is slightly faster by moving expensive computations outside the linting function (#2348, @AshesITR).
```
Comment (MichaelChirico) [inline NEWS.md:26]: may as well cite #2339 too if we're going to mention in NEWS

File: R/backport_linter.R (hunk 1 above)
Comment (AshesITR) [reply R/backport_linter.R:47]: We are using `backport_index` in the code.
I can shorten the code that produces it if you like?

**Review by MichaelChirico — APPROVED** [review]

**Pull request merged**